src/main/java/com/notably/model/block/BlockImpl.java

```diff
@@ -0,0 +1,121 @@
+package com.notably.model.block;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Represents a Block in the Notably data structure.
+ */
+public class BlockImpl implements Block {
+    private List<Block> children;
+    private Title title;
+    private Body body;
+    private Block parent;
+
+    /**
+     * Initializes a default block.
+     * Used when creating a root block and creating a block without the optional body argument.
+     */
+    public BlockImpl(Block parent, Title title) {
+        this.title = title;
+        this.parent = parent;
+    }
+
+    /**
+     * Initializes a block with body content.
+     * Used when creating a block with additional body argument.
+     */
+    public BlockImpl(Block parent, Title title, Body body) {
+        this.title = title;
+        this.parent = parent;
+        this.body = body;
+    }
+
+    @Override
+    public Title getTitle() {
+        return this.title;
+    }
+
+    @Override
+    public Body getBody() {
+        return this.body;
+    }
+
+    @Override
+    public Optional<Block> getParent() {
+        return Optional.ofNullable(parent);
+    }
+
+    @Override
+    public List<Block> getChildren() {
+        return this.children;
+    }
+
+    @Override
+    public void setChildren(List<Block> newChildren) {
+        this.children = newChildren;
+    }
+
+    @Override
+    public Optional<Block> getChild(Title title) {
+        Optional<Block> child = this.children.stream()
+            .filter(block -> title.equals(block.getTitle()))
+            .findAny();
+        return child;
+    }
+
+    @Override
+    public void setChild(Title title, Block newBlock) {
```

Comment:
Let's be a bit defensive and do a nullity check for both `title` and `newBlock`?